### ISSUE TITLE
Only 168.000 of 5.500.000 links found

### ISSUE THREAD

**Issue by Jaaap** [rockdaboot/wget2#100]
When i try to copy a large intranet site, using this command:
`wget2 -owgetlog -4 -m --page-requisites --no-parent --no-use-server-timestamps --content-on-error --max-threads=4 'http://my-intranet-site.tld/'&`

wget2 only finds about 168.000 pages/links where wget found 5.500.000.

Is there a limit to how many links wget2 can hold in memory or am i doing something wrong?

**Comment by rockdaboot** [comment]
I guess, no one made such large tests with wget2 so far :-)

There might be several reasons and I just have to guess without further information.
- Wget2 might obey robots.txt more closely than wget
- Wget2 might be crashed at some point
- Wget2 might have unknown bugs in parsing
- How large is wgetlog ? Maybe there is a 2/4 GB limit ? (32bit or 64bit system on client side ?)

If you are able to find a page that isn't correctly parsed (use -d output with wget2 and wget to compare the found links), please send it to me.

Basically, wget2 is able to find *more* links than wget (support for several HTML5 tags that wget does not support, compressed and uncompressed sitemaps, ...)

**Comment by rockdaboot** [comment]
i also recently made huge refactorings... and it looks like there is something wrong. I have issues downloading huge sites (too many open file descriptors). Let me first check that out before you invest any more time.

**Comment by rockdaboot** [comment]
I found a problem with HTTP/2 - sometimes streams are not reported as 'done' and thus a file descriptor isn't closed. Could you test with --no-http2 ? I would like to know if your problem is a different one or not.

**Comment by Jaaap** [comment]
Ok i will.
The -d run is done and what is surprising (to me) is that it ends with 150.000 of these entries:
```
15.144020.254 blacklist http://mydomain.tld/de/1/E88.html
15.144020.254 blacklist http://mydomain.tld/img/2000003983455.png
15.144020.254 blacklist http://mydomain.tld/de/doc/2000006659756.html
15.144020.254 blacklist http://mydomain.tld/img/2000004047160.png
15.144020.254 blacklist http://mydomain.tld/fr/X/E70/X5_40iX/N55.html
15.144020.254 blacklist http://mydomain.tld/img/2000004068858.png
15.144020.254 blacklist http://mydomain.tld/fr/doc/2000006664836.html
15.144020.254 blacklist http://mydomain.tld/zh/2/F23/225d/B47.html
15.144020.254 blacklist http://mydomain.tld/pt/doc/20000112860843.html
15.144020.254 blacklist http://mydomain.tld/es/doc/2000006723533.html
15.144020.254 blacklist http://mydomain.tld/sv/doc/20000143223356.html
15.144020.254 blacklist http://mydomain.tld/img/2000003953582.png
```
whut? why?

**Comment by rockdaboot** [comment]
Thats all the URLs that have been found & processed. The 'blacklist' is the 'done' list.

**Comment by rockdaboot** [comment]
You can make a list of what files downloaded and which URLs been found in them (and which/why not followed). Use --max-threads=1. When done 
```
egrep 'GET /|*url|not followed' wgetlog
```
This should print out a list... like
```
GET /page6/page6.html HTTP/1.1
15.155727.304 *url = page6.html
15.155727.305 *url = files/page6_1.jpg
```
Hopefully you can check and find why URLs are not followed.
These `*url` lines say what wget2 found in the document from the previous `GET` line.
In case you see wget2 not parsing the/all URLs from a document... please send me that file.

**Comment by rockdaboot** [comment]
BTW, -m (--mirror) implicitely sets --level=INF. I wonder how --page-requisites makes since in this context (the page requisites are already included in infinite recursion). Maybe wget2 behaves different and limits the URLs to download to the page requisites (only follow A and LINK tags). Can you give it try without --page-requisites to see if it makes a difference ?

**Comment by Jaaap** [comment]
- Adding --no-http2 does not make a difference
- Now trying --max-threads=1 -d without --page-requisites (will do the grep thing afterwards)

**Comment by rockdaboot** [comment]
Do you see debug messages like 'cookie '...' dropped, domain '...' is a public suffix' in the output of wget2 ? Just in case...

**Comment by Jaaap** [comment]
- There's nothing like 'cookie '...' dropped, domain '...' is a public suffix' in the log.
- I've mailed you the egrep on the log
- For most pages only the first div with <a/>'s is shown, the rest is skipped (might be in illegal html though, like <ul><a href="fsdf">fdgg</a></ul>)

**Comment by rockdaboot** [comment]
Thank you very much for reporting.

Identical relative URLs in different pages were skipped - a big bug since when I introduced 'known_urls'.

Just pushed a change including a test case.
Please give it a try.
